internal/migration/migration.go

```diff
@@ -0,0 +1,478 @@
+	default:
+		return fmt.Errorf("unknown destination type %q for migration package", destination)
+	}
+}
+
+func shouldMigrateBolt(sourceBeaconPath string) error {
+	existingDB, err := bbolt.Open(sourceBeaconPath, ownerOnly, nil)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = existingDB.Close()
+	}()
+
+	return existingDB.View(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(bucketName)
+		_, value := bucket.Cursor().First()
+		b := beacon{}
+		err := json.Unmarshal(value, &b)
+		if err != nil {
+			return ErrMigrationNotNeeded
+		}
+
+		return nil
+	})
+}
+
+func pgConn(ctx context.Context, logger log.Logger, beaconName, pgDSN string) (*pgdb.Store, func(), error) {
+	pgConf, err := database.ConfigFromDSN(pgDSN)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	conn, err := database.Open(ctx, pgConf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err := schema.Migrate(ctx, conn); err != nil {
+		err := fmt.Errorf("migrating error: %w", err)
+		return nil, nil, err
+	}
+
+	store, err := pgdb.NewStore(ctx, logger, conn, beaconName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cancel := func() {
+		_ = store.Close(ctx)
+		_ = conn.Close()
+	}
+
+	return store, cancel, err
+}
+
+func shouldMigratePostgres(ctx context.Context, logger log.Logger, beaconName, pgDSN string) error {
+	store, cancel, err := pgConn(ctx, logger, beaconName, pgDSN)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	storeLen, err := store.Len(ctx)
+	if err != nil {
+		return err
+	}
+
+	if storeLen != 0 {
+		logger.Warnw("Postgres contains beacon rounds. Skipping migration", "rounds count", storeLen)
+		return ErrMigrationNotNeeded
+	}
+
+	return nil
+}
+
+func (m *migrator) doMigrate(ctx context.Context) error {
+	defer func() {
+		finishedIn := time.Since(m.startedAt).String()
+		m.logger.Infow(
+			"Finished processing beacons",
+			"beaconName", m.beaconName,
+			"finishedIn", finishedIn,
+		)
+	}()
+
+	//nolint:exhaustive // We want to explicitly ignore the chain.MemDB backend since there's nothing to migrate there.
+	switch m.destination {
+	case chain.BoltDB:
+		go m.migrateBolt()
+	case chain.PostgreSQL:
+		go m.migratePostgres(ctx)
+	}
+
+	// Spin up the reader
+	go m.reader()
+
+	return m.wait()
+}
+
+//nolint:funlen // This function has the right length
+func (m *migrator) migratePostgres(ctx context.Context) {
+	rows := 0
+	defer func() {
+		if m.existingRows != rows {
+			err := fmt.Errorf("not all rounds migrated successfully expected: %d actual: %d", m.existingRows, rows)
+			m.errChan <- err
+			return
+		}
+	}()
+	defer func() {
+		finishedIn := time.Since(m.startedAt).String()
+		m.logger.Infow("finished saving data in postgres", "rows", rows, "finishedIn", finishedIn)
+	}()
+	defer m.ctxCancel()
+	defer func() {
+		if m.existingRows != rows {
+			m.logger.Errorw("not all rounds migrated successfully", "expected", m.existingRows, "actually", rows)
+			return
+		}
+	}()
+
```

Comment:
I see several calls to `m.errChan <- err`, where is the receiver for this?

Reply:
How much better would it be if this function returned an error and any concurrency was handled all together in the calling function?